### ISSUE TITLE
Need to set the model to 'eval' mode when assessing the validation data

### ISSUE THREAD

**Issue by danarteaga** [wolny/pytorch-3dunet#53]
FYI - I believe that you need to add self.model.eval() before you run a forward pass of the validation data. This should then be changed back to self.model.train() when calculating loss for the training data again. 

With the way your code is currently configured (using group normalization), this will not affect anything. However, if the user were to either i) perform dropout or ii) use batch normalization, they could run into issues. 

See: 
https://stackoverflow.com/questions/55627780/evaluating-pytorch-models-with-torch-no-grad-vs-model-eval 
and
https://discuss.pytorch.org/t/what-does-model-eval-do-for-batchnorm-layer/7146/3
for more info. 

They way that I came across this is that I used batch normalization (erroneously, for a batch size of 1), and during testing my predictions didn't make any sense. However, when I set the model to 'train' mode and explicitly passed it through the activation layer, the predictions were fine.

**Comment by joaco18** [comment]
Hi @danarteaga ! Where did you added the self.model.eval() and self.model.train() lines? 
I was adding a dropout layer after the bottleneck of the model, so as you said I needed to turn the model to de evaluation mode before the validation steps.

I did it this way: in the trainer.py file in line 185:
                self.model.eval()
                eval_score = self.validate(self.loaders['val'])
                self.model.train()   

However when I added those lines the validation metrics dropped down with respect to the previous run without changing the model mode. I supposed that this was a learning/architecture problem (the dropout wasn´t helping), but then I tried again using the new version of the code with model.eval() and model.train(), but with no dropout, and again the performance of my segmentations drop down in validation. In both cases validation scores and losses are very different (worse) that training ones.

A curious thing is that the predictions were consistent with the metrics obtained during training, so my intuition is that my problem is in the implementation of the transition do eval mode and turning it back to train mode. 

Any help will be appreciated!! Thanks!

**Comment by danarteaga** [comment]
Yes, I should have been more explicit before.

In my code, I added:
>> self.model.eval()
at line 184 in the trainer.py script.

You need to comment out lines 243 and 244 of the trainer.py file as well. If the model is in evaluation mode, then by default, self.training = False and self.final_activation(x) will be run in model.py. Thus, if you were to also add lines 243 and 244, you would be applying the activation layer twice, which is no bueno. 

Lastly, as you stated above, make sure to add back 
>> self.model.train()	
somewhere around line 192 so that the model is set back to training mode after validation is over.

That should hopefully fix your issues.

**Comment by wolny** [comment]
Hi @danarteaga,

nice catch! You're totally right, `model.eval()` should definitely be called during the validation phase. I've fixed the issue, see: 8105abebefcf305c2e4a596e99db44b49280520f

Thanks!
